### ISSUE TITLE
Feature request: ability to re-export a module

### ISSUE THREAD

**Issue by tchajed** [dafny-lang/dafny#1133]
I'd find it really useful to be able to re-export a module. Many of my modules are intended to be used opened and as you go up the stack, import lists just keep growing since you need all the lower-level modules for their types. This is bad for modularity since every module has a long list of modules that have to be `import opened` alongside them to make things usable. I also can't split up a module's implementation into several smaller files without creating a bunch of small modules that the caller will know about (which is just one instance of poor modularity).

Am I doing something wrong or missing something?

**Comment by RustanLeino** [comment]
You are right that the `opened`-ness of imports applies only in the importing module itself. That is, openness is not transferred. This is intentional, so that one module can choose which names to use as un-qualified without imposing the same choice on all its importers.

If it's just a matter of number of characters required in the program text, then Dafny's import mechanism allows you to give a local name for the module you import. For example, if you write

``` dafny
import N = Network
```

then you can refer to the members of the `Network` module with the qualification `N.`. This is not as short as you'd get from `import opened`, but it doesn't merge namespaces like `import opened` does.

If many of the members of a module act on a value of a `class`, `trait`, `datatype`, or `newtype` declared inside the module, then you can choose to move those members from the module-level to the type-level. Thus, instead of writing

``` dafny
var c := Network.CreateConnection();
Network.Send(c, "hello");
```

a client of the module would write

``` dafny
var c := Network.CreateConnection();
c.Send("hello");
```

One might argue (as I would) that `import opened` should be used only sparingly, say for the most common declarations (perhaps like `uint64`), so as not to make namespaces too big. One convention found in C-like programs (that lack qualified names) is that all names of a would-be module start with some prefix, like `network_`. For these, a module system allows one to instead name the module `Network` and to access its members with the prefix `Network.`. I have seen uses of `import opened` with compensating `network_` prefixes of every name. In such a situation, my recommendation is to stop using `import opened` and to replace the self-imposed prefix `network_` prefix with the language-enforced `Network.`.

**Comment by keyboardDrummer** [comment]
@tchajed Can you add a code example of the problem you're facing?

**Comment by tchajed** [comment]
Sure, [this import block](https://github.com/mit-pdos/daisy-nfsd/blob/9c6ad4101ea897b927a4e7d311df058305003c49/src/fs/dir_fs.dfy#L8-L10) shows up in almost every file. One of the libraries I'm importing everywhere ("Machine") defines `byte` and `uint64`, which I don't want to qualify (even with a single-letter module). Another example is ByteSlice, which gives an interface to efficient byte slices in Go.

I think of these libraries as extending the standard library. If I could re-export them, one useful organization would be to implement them as individual components but then gather them up in a `Std` module that re-exports all of them. This way at least there's just a single import.

**Comment by keyboardDrummer** [comment]
@RustanLeino would it be an option to allow exporting members opened from another module by using an export set?

```dafny
module A {
  const b := true
}

module ReExport {
  import opened A
  export provides b
}

module Consume {
  import opened ReExport
  
  method foo() {
    assert b;
  }
}
```

Or to allow exporting members from other modules:

```dafny
module A {
  const b := true
}

module ReExport {
  import A
  export provides A.b
}

module Consume {
  import opened ReExport
  
  method foo() {
    assert b;
  }
}
```

I'm not sure about the right syntax, but at least the concept of exporting members of imported modules seem useful. I don't know how common this is in other languages but TypeScript for example has it: [docs](https://www.typescriptlang.org/docs/handbook/modules.html#re-exports).
